test/workflow_test.go

```diff
@@ -573,6 +575,48 @@ func (w *Workflows) WorkflowWithLocalActivityCtxPropagation(ctx workflow.Context
 	return result, nil
 }
 
+func (w *Workflows) NonDeterminismSimulatorWorkflow(ctx workflow.Context) ([]string, error) {
+	logger := workflow.GetLogger(ctx).Sugar()
+	ctx = workflow.WithActivityOptions(ctx, w.defaultActivityOptions())
+
+	workflow.SetQueryHandler(ctx, "custom_query", func() (int, error) {
+		return w.nonDeterminismSimulatorWorkflowCallCount, nil
+	})
+
+	res := []string{}
+
+	// Mimic non-deterministic behavior with alternating calls to activity here.
+	// Workflow functon is invoked multiple times at every decision point and each time it will behave differently due to this optional activity call.
+	// This will be considered as non-deterministic error
+	if w.nonDeterminismSimulatorWorkflowCallCount%2 == 0 {
+		var ans string
+		if err := workflow.ExecuteActivity(ctx, "Prefix_ToUpper", "hello").Get(ctx, &ans); err != nil {
+			return nil, err
+		}
+		res = append(res, ans)
+	}
+
+	w.nonDeterminismSimulatorWorkflowCallCount++
+
+	selector := workflow.NewSelector(ctx)
+	timer := workflow.NewTimer(ctx, 5*time.Second)
+	selector.AddFuture(timer, func(workflow.Future) {
+		logger.Infof("Timer future is called")
```

Comment:
```suggestion
		logger.Info("Timer future is called")
```
but obviously harmless